buffer/src/main/java/io/netty/buffer/UTF8Decoder.java

```diff
@@ -0,0 +1,210 @@
+            throw new IndexOutOfBoundsException(illegalIndex(offset, length, size));
+        }
+        UTF8Processor processor = new UTF8Processor(new char[length]);
+        buf.forEachByte(offset, length, processor);
+        return processor.toString();
+    }
+
+    public static final class UTF8Processor implements ByteProcessor {
+        private int state = UTF8_ACCEPT; // The state we are currently in, we start out by wanting a new character
+        private int codePoint; // Our current unicode code point
+        private int destLength;
+        private final char[] dest;
+
+        public UTF8Processor(char[] dest) {
+            this.dest = dest;
+        }
+
+        public boolean process(byte rawByte) {
+            int b = rawByte & 0xFF; // Byte.toUnsignedInt
+            if (b <= 127) {
+                processASCII(b);
+            } else {
+                processUnicode(b);
+            }
+            return true;
+        }
+
+        private void processASCII(int b) {
+            dest[destLength++] = (char) b;
+        }
+
+        private void processUnicode(int b) {
+            final int characterClass = CHARACTER_CLASSES[b];
+            if (state == UTF8_ACCEPT) {
+                codePoint = (0xFF >> characterClass) & b;
+            } else {
+                codePoint = (b & 0x3f) | (codePoint << 6);
+            }
+            switch (state = TRANSLATION_TABLE[state + characterClass]) {
+                case UTF8_REJECT:
+                    throw new IllegalArgumentException(illegalByte(b));
+                case UTF8_ACCEPT:
+                    // We're finished
+                    if (isBasicCodePoint(codePoint)) {
+                        // It can be represented as a single char
+                        dest[destLength++] = (char) codePoint;
+                    } else if (isValidCodePoint(codePoint)) {
+                        dest[destLength++] = highSurrogate(codePoint);
+                        dest[destLength++] = lowSurrogate(codePoint);
+                    } else {
+                        throw new IllegalArgumentException(illegalCodePoint(codePoint));
+                    }
+            }
+        }
+
+        public String toString() {
+            return new String(dest, 0, destLength);
+        }
+    }
+
+    private static String illegalByte(int character) {
+        return "Illegal UTF8 byte: 0x " + StringUtil.byteToHexStringPadded(character);
+    }
+
+    private static String illegalCodePoint(int codePoint) {
+        return "Illegal UTF8 code point "
+                + StringUtil.toHexStringPadded(new byte[] {
+                                (byte) (codePoint >> 24),
+                                (byte) (codePoint >> 16),
+                                (byte) (codePoint >> 8),
+                                (byte) codePoint
+                });
+    }
+
+    // Java 6 Backports of utilities in java.lang.Character
+
+    private static boolean isBasicCodePoint(int codePoint) {
```

Comment:
i wonder if these should go into `StringUtil`. We have a `isSurrogate()` which is now provided by the JDK in their.